src/stdatamodels/jwst/datamodels/tests/test_read_metadata.py

```diff
@@ -0,0 +1,91 @@
+"""Test utilities for loading metadata without loading entire model."""
+
+import pytest
+import numpy as np
+from sys import getsizeof
+from astropy.time import Time
+import stdatamodels.jwst.datamodels as dm
+from stdatamodels.jwst.datamodels.util import read_metadata
+
+
+TESTFILE_ROOT = "jwst_image."
+FILT = "GR150R"
+INPUT_TIME = "2021-01-01 00:00:00.000"
+DATA = np.zeros((1000, 1000), dtype=np.float32)
+
+
+@pytest.fixture
+def make_and_save_models(tmp_path):
+    """Create a new datamodel with a meta attribute and relatively large data array."""
+    model = dm.ImageModel(DATA)
+    model.meta.instrument.filter = FILT
+
+    # give this an astropy time to ensure it is cast to string
+    model.meta.observation.date = Time(INPUT_TIME, format="iso", scale="utc")
+
+    for extension in ["fits", "asdf"]:
+        # Save the datamodel to a file
+        root = TESTFILE_ROOT + extension
+        path = str(tmp_path / root)
+        model.save(path)
+        model.close()
+
+
+@pytest.mark.parametrize("extension", ["fits", "asdf"])
+def test_read_metadata(make_and_save_models, extension, tmp_path):
+    """Test metadata loading."""
+    root = TESTFILE_ROOT + extension
+    path = str(tmp_path / root)
+
+    # Load the metadata
+    meta = read_metadata(path)
+
+    # Ensure the dict is flat, i.e., none of the vals in keys are themselves dicts
+    assert isinstance(meta, dict)
+    assert all(not isinstance(v, dict) for v in meta.values())
+
+    # Ensure the metadata has the expected attributes
+    # with times re-cast from astropy Time to string (and not asdf.tagged.TaggedString)
+    assert isinstance(meta["meta.instrument.filter"], str)
+    assert isinstance(meta["meta.observation.date"], str)
+    assert meta["meta.instrument.filter"] == FILT
+    assert meta["meta.observation.date"] == INPUT_TIME
+    assert meta["meta.filename"] == root
+
+    # Ensure attributes in schema but not in model are there but give None
+    assert meta["meta.instrument.detector"] is None
+
+    # Ensure the metadata does not have data attributes
+    assert "data" not in meta
+    assert getsizeof(meta) < getsizeof(DATA)
```

Comment:
This only measures the size of the `meta` dict. [Not the values](https://docs.python.org/3/library/sys.html#sys.getsizeof):
```python
>> d = {'a': 1}
>> sys.getsizeof(d)
184
>> d['a'] = np.zeros((10000, 10000))
>> sys.getsizeof(d)
184
```

Reply:
whoops.

I wanted to use something like `stcal.testing_helpers.MemoryThreshold` but stcal is not a dependency of stdatamodels.  Does it feel like overkill to copy that entire context manager into here just for this test?

Reply:
Probably. What's the goal of the `getsizeof` check? Is it to make sure that data isn't hidden somewhere in the meta dict? Or is the goal to check that `data` was never read during the call to `read_metadata`?

Reply:
The plan was the second one, which I realize was not well thought out - even if the `getsizeof` check did check the memory usage of the entire contents of the dict, it wouldn't have ensured that it was never loaded and then discarded.  The desired use-case is indeed exactly what the `MemoryThreshold` context was built to solve.  But on the other hand, is this test necessary to include at all?

Reply:
It does seem useful to check that `read_metadata` doesn't read the data. I was hoping truncating the files (so they contain the header but impartial data) might work (if `read_metadata` succeeds). However I haven't been able to get that to work.

We could monkeypatch parts of fits/asdf to make sure the data isn't called but that seems brittle. I wonder if we can track filesystem access. I know how to do that for a file object but since the input is a filename/Path I'm not sure how to get that to work. I'll keep poking at this but if we can't come up with a way to test it I think we can go forward with may be a test of memory usage (perhaps using psutil which is already a test dependency).

Reply:
sounds good.  Is there anything wrong with having something similar to that stcal context manager?  It uses `tracemalloc` which is part of the standard library

Reply:
Nothing wrong with that. I doubt psutil provides anything helpful here.

What do you think about this?
```python
def test_no_data(model_path, monkeypatch):
    def throw_error(self):
        raise Exception()

    if 'asdf' in str(model_path):
        import asdf
        monkeypatch.setattr(asdf, "open", throw_error)
        read_metadata(model_path)
    else:
        from astropy.io import fits
        monkeypatch.setattr(fits.ImageHDU, "data", property(throw_error))

        with fits.open(model_path) as hdulist:
            monkeypatch.setattr(fits, "open", lambda *args, **kwargs: hdulist)
            read_metadata(model_path)
```
It's only patching public API so it hopefully shouldn't be too brittle :-/